File: benchmarks/CWE-089-SQLi/safe-parameterized.py
# Safe: Uses parameterized queries with SQLAlchemy
from sqlalchemy import create_engine, text

engine = create_engine('sqlite:///app.db')

def get_user(user_id):
    with engine.connect() as conn:
        # Safe: parameterized with :param syntax
        result = conn.execute(
            text("SELECT * FROM users WHERE id = :id"),
            {"id": user_id}
        )
        return result.fetchone()

def search_products(category, min_price):
    with engine.connect() as conn:
        # Safe: multiple named parameters
        result = conn.execute(
            text("SELECT * FROM products WHERE category = :cat AND price > :price"),
            {"cat": category, "price": min_price}
        )
        return result.fetchall()

File: benchmarks/CWE-089-SQLi/safe-parameterized.js
// Safe: Uses parameterized queries
const mysql = require('mysql2/promise');

async function getUser(userId) {
  const connection = await mysql.createConnection(process.env.DATABASE_URL);
  // Safe: parameterized query with placeholder
  const [rows] = await connection.execute(
    'SELECT * FROM users WHERE id = ?',
    [userId]
  );
  await connection.end();
  return rows[0];
}

async function searchUsers(name, role) {
  const connection = await mysql.createConnection(process.env.DATABASE_URL);
  // Safe: multiple parameters
  const [rows] = await connection.execute(
    'SELECT * FROM users WHERE name LIKE ? AND role = ?',
    [`%${name}%`, role]
  );
  await connection.end();
  return rows;
}
